### PR TITLE
raftstore-v2: fix metrics and perf context

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #12842 

What's Changed:

```commit-message
This PR fixes several bugs and metrics:
- Now waterfall timer will be reset in before_write, the goal is to solve
    the confusion that stall writes can pollute the whole waterfall metrics.
- Perf context is changed not to be associated with engine instance. Perf
      context is thread local and instance independent under the hook.
- Fix flushed index advance failure due to suspicious flush.
- Support print long uncommitted logs and fix incorrect commit time
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/worker/tablet_gc.rs [hunk 1]
```diff
@@ -131,6 +131,7 @@ impl<EK: KvEngine> Runner<EK> {
     }
 
     fn prepare_destroy(&mut self, region_id: u64, tablet: EK, wait_for_persisted: u64) {
+        let _ = tablet.set_db_options(&[("avoid_flush_during_shutdown", "true")]);
```
Comment (tabokie) [inline components/raftstore-v2/src/worker/tablet_gc.rs:134]: Needs comment the direct reason for it.

File: components/raftstore-v2/src/operation/ready/apply_trace.rs [hunk 2]
```diff
@@ -257,10 +258,16 @@ impl ApplyTrace {
...
-            .iter()
+            .iter_mut()
             // Only unflushed CFs are considered. Flushed CF always have uptodate changes
             // persisted.
             .filter_map(|pr| {
+                // All modifications before mem_index must be seen. If following condition is true,
+                // it means the modification comes beyond general apply process (like transaction GC unsafe write). Align `last_modified`
+                // to `flushed` to avoid blocking raft log GC.
+                if mem_index >= pr.flushed && pr.flushed > pr.last_modified {
+                        pr.last_modified = pr.flushed;
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/ready/apply_trace.rs:269]: what's with the indentation.

**Review by tabokie — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

**Review by tabokie — APPROVED** [review]

**Pull request merged**